kafka/metrics.go

```diff
@@ -0,0 +1,120 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kafka
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const instrumentName = "github.com/elastic/apm-queue/kafka"
+const unitCount = "1"
+
+type instruments struct {
+	MessageProduced metric.Int64Counter
+	WriteErrors     metric.Int64Counter
+	WriteTimeout    metric.Int64Counter
+}
+
+type kgoHooks struct {
+	instruments instruments
+}
+
+func NewKgoHooks(mp metric.MeterProvider) *kgoHooks {
+	m := mp.Meter(instrumentName)
+
+	a, err := m.Int64Counter(
+		"message.produced.count",
+		metric.WithDescription("The total number of message produced"),
+		metric.WithUnit(unitCount),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	b, err := m.Int64Counter(
+		"write.error.count",
+		metric.WithDescription("The total number of error occurred on write"),
+		metric.WithUnit(unitCount),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	c, err := m.Int64Counter(
+		"write.timeout.count",
+		metric.WithDescription("The total number of messages not produced due to timeout"),
+		metric.WithUnit(unitCount),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return &kgoHooks{
+		instruments{
+			MessageProduced: a,
+			WriteErrors:     b,
+			WriteTimeout:    c,
```

Comment:
I find it hard to correlate the `a`, `b`, `c` variables with their instruments. Could you change the variable names to resemble the metric they represent?

Reply:
Also, I'd prefer to a name `instrument` field, rather than specifying the `instruments` as the only struct argument.

Reply:
I changed the names to more readable ones